Add methods to work with built images

mkosiarc opened the pull request:
CLOUDBLD-8247

Signed-off-by: mkosiarc <mkosiarc@redhat.com>

# Maintainers will complete the following section

- [x] Commit messages are descriptive enough
- [x] Code coverage from testing does not decrease and new code is covered
- [x] JSON/YAML configuration changes are updated in the relevant schema
- [n/a] Changes to metadata also update the documentation for the metadata
- [n/a] Pull request has a link to an osbs-docs PR for user documentation updates
- [n/a] New feature can be disabled from a configuration file

chmeliik commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +99,98 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+        :returns Dict[str, int], where key is layer id and int is the size
+                                 in bytes
+        """
+        unpack_cmd = f"tar xf {path}"
+        layers_cmd = "jq < manifest.json '.[0].Layers[]' -r | xargs du -b"
```
Coud we do this with python's `tarfile` and `json` modules instead of subprocess calls?

The cool thing is that we probably won't even have to extract the entire archive that way:
```python
with tarfile.open(docker_tarball) as tar:
    manifest = json.load(tar.extractfile("manifest.json"))
    layer_size = tar.getmember(...).size
```
And similarly for the other method

mkosiarc replied:
This does make more sense, I don't know why I reasoned myself out of this approach. updated now

mkosiarc commented on atomic_reactor/inner.py:
```diff
@@ -858,19 +858,10 @@ def build_docker_image(self) -> BuildResult:
-                    for (diff_id, layer) in zip(diff_ids, reversed(history))
-                ]
+                self.data.built_image_inspect = imageutil.get_inspect_for_image(self.data.image_id)
+                image_archive_tarball_path = "" # TODO
```
@chmeliik what would be good way to handle this ? Just create directory in the specific platform build dir ?

chmeliik replied:
I'm thinking the best thing to do might be to remove `built_image_inspect` and `layer_sizes` from WorkflowData entirely. The few plugins that need this info can fetch it using imageutil methods instead of taking it from context data.

If it would be too much trouble to remove at this point, maybe just don't set them here and add `# OSBS2 TBD`s about removing them to the affected plugins

mkosiarc commented:
Do you have idea what to do with this? 
`atomic_reactor/utils/imageutil.py:140: error: Argument 1 to "load" has incompatible type "Optional[IO[bytes]]"; expected "SupportsRead[Union[str, bytes]]"`

It didn't help if I forced it to be str

mkosiarc commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,71 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+            layer_file = tar.getmember(layers[0])
+            tar.extract(layer_file, dst_path)
 
-def get_image(image):
```
@acmiel post_compress.py calls this method, is `download_image_archive_tarball` doing the same? or should this method do something different?

chmeliik replied:
I think the only difference is that `download_image_archive_tarball` creates a file on disk while the current usage assumes that `get_image` returns an open binary file. We don't need to keep `get_image` for that, just open the downloaded tarball instead

mkosiarc replied:
Thanks, I updated the todo comment in post_compress.py and removed the call to get_image

mkosiarc commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,71 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+        :param src_path: str, path to image archive tarball
+        :param dst_path: str, path where the layer will be copied
+
+        :returns str, the filesystem layer name
```
It does, see ```return layers[0]```, is that just github diff shows it below the removed get_image method

mkosiarc commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,71 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
-    # most likely won't be needed at all
-    return {}
+        try:
+            return subprocess.check_output(cmd, stderr=subprocess.STDOUT)
```
updated, to use retries. That's the right solution here.

mkosiarc commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,71 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+
+        :param path: str path to a image archive tarball
+        :returns Dict[str, int], where key is layer id and int is the size
+                                 in bytes
```
updated

chmeliik commented:
> Do you have idea what to do with this? `atomic_reactor/utils/imageutil.py:140: error: Argument 1 to "load" has incompatible type "Optional[IO[bytes]]"; expected "SupportsRead[Union[str, bytes]]"`
> 
> It didn't help if I forced it to be str

I'm guessing you need to get rid of the `Optional` in `Optional[IO[bytes]]`. For example like this:
```python
def does_typecheck() -> dict:
    with tarfile.open("./archive.tar") as tf:
        manifest_file = tf.extractfile("manifest.json")
        if not manifest_file:
            raise ValueError("manifest.json is not a file")
        manifest = json.load(manifest_file)

    return manifest
```
[tarfile.extractfile](https://docs.python.org/3/library/tarfile.html#tarfile.TarFile.extractfile) returns None if the member is not a regular file and raises KeyError if the member is not present at all

chmeliik commented:
I've been playing around with `oc image extract` a bit locally, and some of the behaviors are potentially surprising (tested with oc version 4.9.0).

#### 1. the dst_path must be an existing empty dir, nothing else works
<details>
<summary>Example</summary>

```shell
$ oc image extract docker.io/library/python:3.10.0-alpine --path /usr/local/bin/pip:/tmp
error: directory /tmp must be empty, pass --confirm to overwrite contents of directory

$ oc image extract docker.io/library/python:3.10.0-alpine --path /usr/local/bin/pip:/tmp/output
error: destination path does not exist: /tmp/output

$ mkdir /tmp/output
$ oc image extract docker.io/library/python:3.10.0-alpine --path /usr/local/bin/pip:/tmp/output
$ ls /tmp/output
pip
```
</details>

#### 2. file permissions are not preserved
<details>
<summary>Example</summary>

```shell
$ podman run --rm -ti docker.io/library/python:3.10.0-alpine sh
/ # stat -c '%A %U:%G' /usr/local/bin/pip
-rwxr-xr-x root:root
/ # exit

$ stat -c '%A %U:%G' /tmp/output/pip 
-rw-r----- acmiel:acmiel
```
</details>

#### 3. nonexistent files are silently skipped
<details>
<summary>Example</summary>

```shell
$ oc image extract docker.io/library/python:3.10.0-alpine --path /usr/bin/pip:/tmp/output
$ echo $?              
0

$ ls /tmp/output -a
.  ..
```
</details>

#### 4. different behavior for `--path /src/dir` vs `--path /src/dir/`
<details>
<summary>Example</summary>

```shell
$ oc image extract docker.io/library/python:3.10.0-alpine --path /usr/local/bin:/tmp/output
$ ls /tmp/output
bin
$ ls /tmp/output/bin
2to3       idle3     pip3     pydoc3     python3            python3-config
2to3-3.10  idle3.10  pip3.10  pydoc3.10  python3.10         python-config
idle       pip       pydoc    python     python3.10-config  wheel
```
```shell
$ oc image extract docker.io/library/python:3.10.0-alpine --path /usr/local/bin/:/tmp/output
$ ls /tmp/output
2to3       idle3     pip3     pydoc3     python3            python3-config
2to3-3.10  idle3.10  pip3.10  pydoc3.10  python3.10         python-config
idle       pip       pydoc    python     python3.10-config  wheel
```
</details>

They don't seem too bad, the limitation of not preserving permissions was already known and we can live with it, the rest is just minor traps for unsuspecting devs. It would be good to document them though.

chmeliik commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+            if not manifest_file:
+                raise ValueError(f'manifest.json from {path} is not a regular file')
+            manifest = json.load(manifest_file)
+            layers = manifest[0]['Layers']
```
Should we also check that there is exactly one item in the manifest? The docker [spec](https://github.com/moby/moby/blob/master/image/spec/v1.2.md#combined-image-json--filesystem-changeset-format) says there might optionally be entries for the parent images, unfortunately it does not say if the current image will be first or last in the list

mkosiarc replied:
I am not sure about this one, from quick look I can't really find anything about how to identify the item we want when there is more items. I'll try to look into it some more

tkdchen replied:
It would be also good to separate these lines into a function:

```python
        with tarfile.open(src_path) as tar:
            manifest_file = tar.extractfile('manifest.json')
            if not manifest_file:
                raise ValueError(f'manifest.json from {src_path} is not a regular file')
            manifest = json.load(manifest_file)
            layers = manifest[0]['Layers']
```

or in some other ways in order to not repeat similar logic in two methods.

mkosiarc replied:
@tkdchen I am not sure separate method is necessary, especially know that one function also extracts config.json, but If you feel strongly about this I'll refactor this.

mkosiarc replied:
I didn't find any additional info about having the entries for parent images, so I am not sure how to proceed. I am assuming that our images will have only one entry (I was testing with couple of them), but to avoid any surprises in the future, maybe I could document this in the code and also raise an error?

chmeliik replied:
document and raise error sounds good :+1:

mkosiarc replied:
@chmeliik updated

chmeliik commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+            layers = manifest[0]['Layers']
+            layer_sizes = []
+            for layer in layers:
+                layer_id = f'sha256:{layer.replace(".tar", "")}'
```
Based on the above-mentioned spec, it seems that the file names of layers should not be relied on. The fact that they match the diff_id seems to be a skopeo implementation detail.

Instead, we could get the diff IDs from `rootfs.diff_ids` in the config file (also present in the docker archive and referenced in manifest.json) and rely on the fact that the `rootfs.diff_ids` array and the `Layers` array are in the same order.

mkosiarc replied:
updated

chmeliik commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+
+    def extract_filesystem_layer(self, src_path: str, dst_path: str) -> str:
+        """Extract filesystem layer from image archive tarball and
+        saves it at dst_path. This is meant for flatpacks and will work
```
typo: flatpaks, not flatpacks

(more like flatpak is the typo, but an intentional one :smile:)

mkosiarc replied:
fixed

chmeliik commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+
+        :param src_path: str, path to image archive tarball
+        :param dst_path: str, path where the layer will be copied
+        :return: str, the filesystem layer name
```
Technically doesn't have to be just a name, it can be a relative path (example from the spec: `a65da33792c5187473faa80fa3e1b975acba06712852d1dea860692ccddf3198/layer.tar`)

Shouldn't be a problem though, the usage is still the same (treat the return value of this method as a relative path within dst_path)

mkosiarc replied:
updated the docstring slightly

chmeliik commented on tests/utils/test_imageutil.py:
```diff
@@ -37,6 +43,18 @@ def test_inspectable(image, is_inspectable):
     assert imageutil.image_is_inspectable(image) == is_inspectable
 
 
+def mock_tarball(tarball_path, layer_files, manifest_file_content) -> str:
```
Does not return anything

mkosiarc replied:
fixed

chmeliik commented on atomic_reactor/schemas/workflow_data.json:
```diff
@@ -33,11 +33,6 @@
-    "built_image_inspect": {"type": ["object", "null"]},
-    "layer_sizes": {
-      "type": "array",
-      "items": {"$ref": "#/definitions/layer_size"}
```
Can we also remove the layer_size definition now, or is it still used?

mkosiarc replied:
removed

chmeliik commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
+        """Returns data about the uncompressed image layer sizes
+
+        :param path: str path to a image archive tarball
+        :return: List[Dict[str, object]], where key is layer id and int is the size
+                                 in bytes
```
If it's not possible to further specify the value type, it should be `Any` rather than `object`. Not sure if there are any practical differences, but `Any` is the idiomatic way to leave the type unspecified.

And the description of the return value appears outdated

mkosiarc replied:
fixed

mkosiarc commented on atomic_reactor/utils/imageutil.py:
```diff
@@ -92,33 +100,76 @@ def _get_registry_client(self, registry: str) -> util.RegistryClient:
         session = util.RegistrySession.create_from_config(self._conf, registry)
         return util.RegistryClient(session)
 
+    def extract_file_from_image(self, image: str, src_path: str, dst_path: str) -> None:
```
fixed

chmeliik reviewed (commented):
Mostly LGTM, just that one unresolved comment + it would be nice to mention the points from https://github.com/containerbuildsystem/atomic-reactor/pull/1810#issuecomment-1046759412 in the docstring

mkosiarc commented:
> Mostly LGTM, just that one unresolved comment + it would be nice to mention the points from [#1810 (comment)](https://github.com/containerbuildsystem/atomic-reactor/pull/1810#issuecomment-1046759412) in the docstring

I updated the docstring, and also added some checks that might catch some of the issues.

chmeliik reviewed (commented):
LGTM :taco: 

(just the one "document and raise error" point left)
